File: packages/yakr-core/src/yakr_core/invite.py
from __future__ import annotations

import base64
import hashlib
import secrets
import time
from dataclasses import dataclass

import cbor2
from cryptography.hazmat.primitives.asymmetric import ed25519

from yakr_core.hybrid_pq import HYBRID_PQ_CAPABILITY
from yakr_core.identity import Identity, b64decode, b64encode


PROTOCOL_V4 = "yakr-v0.4"
PROTOCOL_V6 = "yakr-v0.6"
SUPPORTED_PROTOCOLS = {PROTOCOL_V4, PROTOCOL_V6}
DEFAULT_INVITE_TTL_MS = 24 * 60 * 60 * 1000


@dataclass(frozen=True)
class InviteBundle:
    protocol: str
    inviter_name: str
    signing_public: bytes
    agreement_public: bytes
    invite_secret: bytes
    rendezvous_hint: str
    expires_at: int
    capabilities: tuple[str, ...]
    signature: bytes
    kem_public: bytes = b""
    pq_signing_public: bytes = b""
    pq_signature: bytes = b""

    def _unsigned_dict(self) -> dict[str, object]:
        payload: dict[str, object] = {
            "protocol": self.protocol,
            "inviter_name": self.inviter_name,
            "signing_public": self.signing_public,
            "agreement_public": self.agreement_public,
            "invite_secret": self.invite_secret,
            "rendezvous_hint": self.rendezvous_hint,
            "expires_at": self.expires_at,
            "capabilities": list(self.capabilities),
        }
        if self.kem_public:
            payload["kem_public"] = self.kem_public
        if self.pq_signing_public:
            payload["pq_signing_public"] = self.pq_signing_public
        return payload

    def unsigned_payload(self) -> bytes:
        return cbor2.dumps(self._unsigned_dict())

    def to_bytes(self) -> bytes:
        payload = self._unsigned_dict()
        payload["signature"] = self.signature
        if self.pq_signature:
            payload["pq_signature"] = self.pq_signature
        return cbor2.dumps(payload)

    @classmethod
    def from_bytes(cls, data: bytes) -> InviteBundle:
        payload = cbor2.loads(data)
        return cls(
            protocol=str(payload["protocol"]),
            inviter_name=str(payload["inviter_name"]),
            signing_public=bytes(payload["signing_public"]),
            agreement_public=bytes(payload["agreement_public"]),
            invite_secret=bytes(payload["invite_secret"]),
            rendezvous_hint=str(payload["rendezvous_hint"]),
            expires_at=int(payload["expires_at"]),
            capabilities=tuple(str(item) for item in payload["capabilities"]),
            signature=bytes(payload["signature"]),
            kem_public=bytes(payload.get("kem_public", b"")),
            pq_signing_public=bytes(payload.get("pq_signing_public", b"")),
            pq_signature=bytes(payload.get("pq_signature", b"")),
        )


def invite_supports_hybrid(bundle: InviteBundle) -> bool:
    return HYBRID_PQ_CAPABILITY in bundle.capabilities and bool(bundle.kem_public)


def create_invite(
    identity: Identity,
    *,
    rendezvous_hint: str,
    ttl_ms: int = DEFAULT_INVITE_TTL_MS,
    hybrid_pq: bool = False,
) -> InviteBundle:
    capabilities = ["direct_p2p", "friend_relay", "store_forward"]
    protocol = PROTOCOL_V4
    kem_public = b""
    pq_signing_public = b""
    pq_signature = b""

    if hybrid_pq:
        if not identity.kem_public:
            raise ValueError("identity missing ML-KEM keypair for hybrid invite")
        protocol = PROTOCOL_V6
        capabilities.append(HYBRID_PQ_CAPABILITY)
        kem_public = identity.kem_public
        if identity.pq_signing_private:
            from pqcrypto.sign.ml_dsa_65 import sign as pq_sign

            pq_signing_public = identity.pq_signing_public
            unsigned = {
                "protocol": protocol,
                "inviter_name": identity.name,
                "signing_public": identity.signing_public_bytes,
                "agreement_public": identity.agreement_public_bytes,
                "invite_secret": secrets.token_bytes(32),
                "rendezvous_hint": rendezvous_hint,
                "expires_at": int(time.time() * 1000) + ttl_ms,
                "capabilities": capabilities,
                "kem_public": kem_public,
                "pq_signing_public": pq_signing_public,
            }
            payload = cbor2.dumps(unsigned)
            signature = identity.signing_private.sign(payload)
            pq_signature = pq_sign(identity.pq_signing_private, payload)
            return InviteBundle(
                protocol=unsigned["protocol"],
                inviter_name=unsigned["inviter_name"],
                signing_public=unsigned["signing_public"],
                agreement_public=unsigned["agreement_public"],
                invite_secret=unsigned["invite_secret"],
                rendezvous_hint=unsigned["rendezvous_hint"],
                expires_at=unsigned["expires_at"],
                capabilities=tuple(unsigned["capabilities"]),
                signature=bytes(signature),
                kem_public=kem_public,
                pq_signing_public=pq_signing_public,
                pq_signature=pq_signature,
            )

    unsigned = {
        "protocol": protocol,
        "inviter_name": identity.name,
        "signing_public": identity.signing_public_bytes,
        "agreement_public": identity.agreement_public_bytes,
        "invite_secret": secrets.token_bytes(32),
        "rendezvous_hint": rendezvous_hint,
        "expires_at": int(time.time() * 1000) + ttl_ms,
        "capabilities": capabilities,
    }
    payload = cbor2.dumps(unsigned)
    signature = identity.signing_private.sign(payload)
    return InviteBundle(
        protocol=unsigned["protocol"],
        inviter_name=unsigned["inviter_name"],
        signing_public=unsigned["signing_public"],
        agreement_public=unsigned["agreement_public"],
        invite_secret=unsigned["invite_secret"],
        rendezvous_hint=unsigned["rendezvous_hint"],
        expires_at=unsigned["expires_at"],
        capabilities=tuple(unsigned["capabilities"]),
        signature=bytes(signature),
        kem_public=kem_public,
        pq_signing_public=pq_signing_public,
        pq_signature=pq_signature,
    )


def verify_invite(bundle: InviteBundle) -> None:
    if bundle.protocol not in SUPPORTED_PROTOCOLS:
        raise ValueError("unsupported invite protocol")
    if bundle.expires_at <= int(time.time() * 1000):
        raise ValueError("invite expired")
    public_key = ed25519.Ed25519PublicKey.from_public_bytes(bundle.signing_public)
    public_key.verify(bundle.signature, bundle.unsigned_payload())
    if bundle.pq_signature:
        if not bundle.pq_signing_public:
            raise ValueError("invite missing PQ signing public key")
        from pqcrypto.sign.ml_dsa_65 import verify as pq_verify

        if not pq_verify(bundle.pq_signing_public, bundle.unsigned_payload(), bundle.pq_signature):
            raise ValueError("invalid PQ invite signature")


def invite_to_url(bundle: InviteBundle) -> str:
    encoded = base64.urlsafe_b64encode(bundle.to_bytes()).decode("ascii").rstrip("=")
    return f"yakr://invite/{encoded}"


def invite_from_url(url: str) -> InviteBundle:
    prefix = "yakr://invite/"
    if not url.startswith(prefix):
        raise ValueError("invalid invite url")
    return InviteBundle.from_bytes(b64decode(url[len(prefix) :]))


def safety_code(bundle: InviteBundle) -> str:
    digest = hashlib.sha256(bundle.signing_public + bundle.agreement_public).digest()
    digits = "".join(str(byte % 10) for byte in digest[:10])
    return f"{digits[0:4]} {digits[4:8]} {digits[8:10]}"

File: rust/yakr-core/src/invite.rs
use std::time::{SystemTime, UNIX_EPOCH};

use ed25519_dalek::Signer;
use rand::RngCore;

use crate::hybrid_pq::HYBRID_PQ_CAPABILITY;
use crate::identity::Identity;
use crate::message::{b64decode, b64encode};

pub const PROTOCOL_V4: &str = "yakr-v0.4";
pub const PROTOCOL_V6: &str = "yakr-v0.6";
pub const DEFAULT_INVITE_TTL_MS: u64 = 24 * 60 * 60 * 1000;

#[derive(Debug, Clone)]
pub struct InviteBundle {
    pub protocol: String,
    pub inviter_name: String,
    pub signing_public: [u8; 32],
    pub agreement_public: [u8; 32],
    pub invite_secret: [u8; 32],
    pub rendezvous_hint: String,
    pub expires_at: u64,
    pub capabilities: Vec<String>,
    pub signature: Vec<u8>,
    pub kem_public: Vec<u8>,
}

pub fn invite_supports_hybrid(bundle: &InviteBundle) -> bool {
    bundle.capabilities.iter().any(|c| c == HYBRID_PQ_CAPABILITY) && !bundle.kem_public.is_empty()
}

pub fn create_invite(
    identity: &Identity,
    rendezvous_hint: &str,
    ttl_ms: u64,
    hybrid_pq: bool,
) -> Result<InviteBundle, String> {
    let mut capabilities = vec![
        "direct_p2p".into(),
        "friend_relay".into(),
        "store_forward".into(),
    ];
    let mut protocol = PROTOCOL_V4.to_string();
    let mut kem_public = Vec::new();
    if hybrid_pq {
        if identity.kem_public.is_empty() {
            return Err("identity missing ML-KEM keypair for hybrid invite".into());
        }
        protocol = PROTOCOL_V6.to_string();
        capabilities.push(HYBRID_PQ_CAPABILITY.into());
        kem_public = identity.kem_public.clone();
    }
    let mut invite_secret = [0u8; 32];
    rand::thread_rng().fill_bytes(&mut invite_secret);
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64;
    let mut bundle = InviteBundle {
        protocol,
        inviter_name: identity.name.clone(),
        signing_public: identity.signing_public_bytes(),
        agreement_public: identity.agreement_public_bytes(),
        invite_secret,
        rendezvous_hint: rendezvous_hint.to_string(),
        expires_at: now + ttl_ms,
        capabilities,
        signature: Vec::new(),
        kem_public,
    };
    let unsigned = invite_unsigned_cbor(&bundle)?;
    bundle.signature = identity.signing_key().sign(&unsigned).to_bytes().to_vec();
    Ok(bundle)
}

pub fn verify_invite(bundle: &InviteBundle) -> Result<(), String> {
    if !yakr_crypto::verify_invite_bundle(
        &invite_to_b64(bundle)?,
        &hex::encode(bundle.signing_public),
        &safety_code(&bundle.signing_public, &bundle.agreement_public),
    ) {
        return Err("invite verification failed".into());
    }
    Ok(())
}

pub fn safety_code(signing_public: &[u8], agreement_public: &[u8]) -> String {
    yakr_crypto::derive_safety_code(signing_public, agreement_public)
}

pub fn invite_to_b64(bundle: &InviteBundle) -> Result<String, String> {
    Ok(b64encode(&invite_to_bytes(bundle)?))
}

pub fn invite_from_b64(value: &str) -> Result<InviteBundle, String> {
    invite_from_bytes(&b64decode(value).map_err(|e| e.to_string())?)
}

fn invite_unsigned_cbor(bundle: &InviteBundle) -> Result<Vec<u8>, String> {
    use ciborium::value::Value;
    yakr_crypto::cbor::encode_cbor(&Value::Map(invite_unsigned_map(bundle)?))
        .map_err(|e| format!("{e:?}"))
}

pub fn invite_to_bytes(bundle: &InviteBundle) -> Result<Vec<u8>, String> {
    use ciborium::value::Value;
    let mut map = invite_unsigned_map(bundle)?;
    map.push((
        Value::Text("signature".into()),
        Value::Bytes(bundle.signature.clone()),
    ));
    yakr_crypto::cbor::encode_cbor(&Value::Map(map)).map_err(|e| format!("{e:?}"))
}

fn invite_unsigned_map(bundle: &InviteBundle) -> Result<Vec<(ciborium::value::Value, ciborium::value::Value)>, String> {
    use ciborium::value::Value;
    let mut entries = vec![
        (Value::Text("protocol".into()), Value::Text(bundle.protocol.clone())),
        (Value::Text("inviter_name".into()), Value::Text(bundle.inviter_name.clone())),
        (Value::Text("signing_public".into()), Value::Bytes(bundle.signing_public.to_vec())),
        (Value::Text("agreement_public".into()), Value::Bytes(bundle.agreement_public.to_vec())),
        (Value::Text("invite_secret".into()), Value::Bytes(bundle.invite_secret.to_vec())),
        (Value::Text("rendezvous_hint".into()), Value::Text(bundle.rendezvous_hint.clone())),
        (Value::Text("expires_at".into()), Value::Integer((bundle.expires_at as i64).into())),
    ];
    let caps: Vec<Value> = bundle
        .capabilities
        .iter()
        .map(|c| Value::Text(c.clone()))
        .collect();
    entries.push((Value::Text("capabilities".into()), Value::Array(caps)));
    if !bundle.kem_public.is_empty() {
        entries.push((
            Value::Text("kem_public".into()),
            Value::Bytes(bundle.kem_public.clone()),
        ));
    }
    Ok(entries)
}

pub fn invite_from_bytes(data: &[u8]) -> Result<InviteBundle, String> {
    let value = yakr_crypto::cbor::decode_cbor(data).map_err(|e| format!("{e:?}"))?;
    let map = value.as_map().ok_or("invite not a map")?;
    let get_bytes = |key: &str| -> Result<Vec<u8>, String> {
        yakr_crypto::cbor::map_bytes(&value, key).ok_or_else(|| format!("missing {key}"))
    };
    let get_text = |key: &str| -> Result<String, String> {
        match yakr_crypto::cbor::map_field(&value, key) {
            Some(ciborium::value::Value::Text(s)) => Ok(s.clone()),
            _ => Err(format!("missing {key}")),
        }
    };
    let caps = yakr_crypto::cbor::map_field(&value, "capabilities")
        .and_then(|v| v.as_array())
        .map(|arr| {
            arr.iter()
                .filter_map(|v| match v {
                    ciborium::value::Value::Text(s) => Some(s.clone()),
                    _ => None,
                })
                .collect()
        })
        .unwrap_or_default();
    Ok(InviteBundle {
        protocol: get_text("protocol")?,
        inviter_name: get_text("inviter_name")?,
        signing_public: get_bytes("signing_public")?.try_into().map_err(|_| "signing_public")?,
        agreement_public: get_bytes("agreement_public")?
            .try_into()
            .map_err(|_| "agreement_public")?,
        invite_secret: get_bytes("invite_secret")?.try_into().map_err(|_| "invite_secret")?,
        rendezvous_hint: get_text("rendezvous_hint")?,
        expires_at: match yakr_crypto::cbor::map_field(&value, "expires_at") {
            Some(ciborium::value::Value::Integer(v)) => i128::from(*v) as u64,
            _ => return Err("missing expires_at".into()),
        },
        capabilities: caps,
        signature: get_bytes("signature")?,
        kem_public: yakr_crypto::cbor::map_bytes(&value, "kem_public").unwrap_or_default(),
    })
}
